okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/UrlSelectorImpl.java

```diff
@@ -41,10 +44,12 @@
 
     private final Supplier<List<HttpUrl>> baseUrls;
     private final AtomicInteger currentUrl;
-    private final Cache<HttpUrl, UrlAvailability> failedUrls;
+    private final LoadingCache<Integer, Instant> failedUrls;
```

Comment:
Doesn't the `List<HttpUrl>` get reshuffled (L57)? using indexes here would mean they end up pointing at something else after the resuffle.

Reply:
You're right, although this creates potential race conditions at multiple places in this class where we compute a URL index and then get the URL using potentially different orderings.

It seems to me that instead of randomizing the URL list we should just randomly select a new current URL every 10 minutes. Would you be open to that change in a follow up?

Reply:
I'd like to see a follow-up that saves the base url into the request as a tag, so we wouldn't have to do any of this indexOf nonsense at all.